### PR TITLE
Optimizations in tensor.py

### PR DESCRIPTION
**Pull request by rayanht**
Using LLaMa for benchmark, `python3 -O` to disable asserts
```
GPU=1 DEBUG=0 OPTLOCAL=2 python3 -O examples/llama.py --prompt "Hello." --count 10 --temperature 0 --timing --profile
```

Aggregating by time spent per file

Before:

```
lazy.py: 0.9999 seconds
tensor.py: 0.8685 seconds
shapetracker.py: 0.4519 seconds
ops.py: 0.3123 seconds
symbolic.py: 0.1109 seconds
```

After:

```
lazy.py: 0.9035 seconds
tensor.py: 0.4017 seconds
shapetracker.py: 0.3556 seconds
ops.py: 0.3054 seconds
symbolic.py: 0.1087 seconds
```

Save 0.43s in `tensor.py`

### CODE REVIEW

**Comment by rayanht** [comment]
Not sure why it's skipping some tests

**Comment by geohot** [comment]
Manually told them to run, yea idk why. Maybe I can pay GitHub more money.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -1,34 +1,67 @@
...
-from typing import List, Tuple, Callable, Optional, ClassVar, Type, Union, Sequence
-from tinygrad.helpers import prod, argfix, make_pair, getenv, IMAGE, DEBUG, flatten, DType, dtypes, ImageDType
+from typing import List, Tuple, Callable, Optional, ClassVar, Type, Union, Sequence, cast
+from tinygrad.helpers import argfix, make_pair, getenv, IMAGE, DEBUG, flatten, DType, dtypes
+from math import ceil, pi, prod, sqrt
 from tinygrad.lazy import Device, LazyBuffer
 from tinygrad.ops import LoadOps
 
 # An instantiation of the Function is the Context
 class Function:
```
Comment (geohot) [inline tinygrad/tensor.py:15]: Hmm, I don't like this split into two classes. This is a lot a copy and paste. And it should support N arguments for CUSTOM functions. How much speed is this gaining?

**Comment by geohot** [comment]
Everything looks fine except the split of the `Function` class. There has to be a cleaner way to do this.

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/mlops.py                         22              22
tinygrad/tensor.py                       113              66
------------------------------------------------------------
total                                    135              88
------------------------------------------------------------
lines added in the tinygrad folder: 47
```

**Comment by rayanht** [comment]
So the rationale for this change is that these two lines in `__init__`, as well as the apply method, completely dominate in the original Function class by trying to accommodate for any number of arguments

```
self.needs_input_grad = [t.requires_grad for t in self.parents]
    self.requires_grad = True if any(self.needs_input_grad) else (None if any(x is None for x in self.needs_input_grad) else False)
```


Before:

```
   ncalls  tottime  percall  cumtime  filename:lineno(function)
    85288    0.147    0.000    2.724  tensor.py:20(apply) <<<<<<<<< Function apply
    88424    0.104    0.000    0.222  tensor.py:37(__init__)
    85288    0.097    0.000    0.154  tensor.py:12(__init__) <<<<<<<< Function __init__
    35424    0.057    0.000    0.919  tensor.py:223(reshape)
    12120    0.052    0.000    1.894  tensor.py:520(_broadcasted)
    24288    0.044    0.000    0.561  tensor.py:227(expand)
```


0.097 + 0.147 = 0.244 seconds spent in the `Function` class

After:

```
   ncalls  tottime  percall  cumtime  filename:lineno(function)
 37080    0.056    0.000    1.566    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:33(apply) <<<<<<<<< Unary apply
    52336    0.049    0.000    0.124    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:70(__init__)
    12120    0.045    0.000    1.277    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:555(_broadcasted)
     2064    0.029    0.000    0.160    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:293(__getitem__)
    16616    0.023    0.000    0.693    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:258(reshape)
   184336    0.021    0.000    0.021    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:112(shape)
    12120    0.020    0.000    0.689    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:49(apply) <<<<<<<<< Binary apply
    12120    0.014    0.000    0.020    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:44(__init__) <<<<<<<<< Binary init
    37080    0.014    0.000    0.014    0.000 /Users/rayanht/tinygrad/tinygrad/tensor.py:28(__init__) <<<<<<<<< Unary init
```

0.056 + 0.014 = 0.07 seconds spent in the `UnaryFunction` class
0.020 + 0.014 = 0.034 seconds spent in the `BinaryFunction` class

Total: 0.041 seconds

A pretty significant gain

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -1,34 +1,67 @@
...
+    return ret
+  
   def forward(self, *args, **kwargs): raise NotImplementedError(f"forward not implemented for {type(self)}")
   def backward(self, *args, **kwargs): raise RuntimeError(f"backward not implemented for {type(self)}")
 
+class BinaryFunction(Function):
+  def __init__(self, device:str, a:Tensor, b:Tensor):
+    self.device, self.parents = device, (a, b)
+    self.needs_input_grad = (a.requires_grad, b.requires_grad)
+    self.requires_grad = True if any(self.needs_input_grad) else (None if any([x is None for x in self.needs_input_grad]) else False)
```
Comment (roelofvandijk) [inline tinygrad/tensor.py:47]: What about `self.requires_grad = True if True in self.needs_input_grad else None if None in self.needs_input_grad else False`? Should be faster.

**Comment by geohot** [comment]
Can you remove that for now? Even with the gains, it violates the tinygrad ethos. We have removed the line count restriction, but the spirit of it remains, and copying and pasting a class with minor changes isn't really worth it for a bit of perf.

**Comment by rayanht** [comment]
Reverted the split, managed to get some performance back thanks to @roelofvandijk's suggestion

Before:

```
lazy.py: 0.9999 seconds
tensor.py: 0.8685 seconds
shapetracker.py: 0.4519 seconds
ops.py: 0.3123 seconds
symbolic.py: 0.1109 seconds
helpers.py: 0.0828 seconds
mlops.py: 0.0614 seconds
linearizer.py: 0.0569 seconds
```

After:

```
lazy.py: 0.8802 seconds
tensor.py: 0.4521 seconds
shapetracker.py: 0.3433 seconds
ops.py: 0.2896 seconds
symbolic.py: 0.1079 seconds
helpers.py: 0.0580 seconds
linearizer.py: 0.0495 seconds
mlops.py: 0.0394 seconds
```

Gain of 0.11s in `shapetracker.py`, 0.4s in `tensor.py` and 0.11s in `lazy.py`

The gains in shapetracker and lazy are indirect from the optimization in `_broadcasted` because we now skip some unnecessary function calls

**Comment by geohot** [comment]
203.27 ms -> 171.39 ms -> 152.18 ms

**Pull request merged**